src/iio_wrapper.cpp

```diff
@@ -26,24 +26,90 @@
 #include <cstdlib>
 #include <sstream>
 #include <rclcpp/rclcpp.hpp>
+#include <unistd.h>
 
-struct iio_context* IIOWrapper::m_network_context = nullptr;
+struct iio_context* IIOWrapper::m_local_context = nullptr;
+
+struct iio_device* IIOWrapper::m_dev = nullptr;
+struct iio_device* IIOWrapper::m_devtrigger = nullptr;
+struct iio_channel** IIOWrapper::m_channels = nullptr;
+struct iio_buffer* IIOWrapper::m_device_buffer = nullptr;
+int IIOWrapper::m_channel_count = 0;
+
+std::mutex IIOWrapper::m_mutex;
 
 IIOWrapper::IIOWrapper()
 {
-    if(m_network_context == nullptr)
+    if(m_local_context == nullptr)
+    {
+        m_local_context = iio_create_local_context();
+        if(!m_local_context)
+        {
+            throw std::runtime_error("Exception: local context is null");
+        }
+    }
+
+    if(m_dev == nullptr)
     {
-        std::string uri = "ip:127.0.0.1"; // TODO: pass uri from cmd line
+        m_dev = iio_context_find_device(m_local_context, "adis16505");
+        if(!m_dev)
+        {
+            throw std::runtime_error("Exception: device data is null");
+        }
+    }
 
-        m_network_context = iio_create_context_from_uri(uri.c_str());
+    if(m_devtrigger == nullptr)
+    {
+        m_devtrigger = iio_context_find_device(m_local_context, "adis16505-dev0");
+        if(!m_devtrigger)
+        {
+            throw std::runtime_error("Exception: device trigger data is null");
+        }
+    }
 
-        m_object_context =  iio_context_clone(m_network_context);
+    if(m_channels == nullptr)
+    {
+        m_channel_count = 0;
+
+        for (int i = 0; i < iio_device_get_channels_count(m_dev); ++i) {
+            struct iio_channel *chn = iio_device_get_channel(m_dev, i);
+            if (iio_channel_is_scan_element(chn))
+            {
+                m_channel_count++;
+            }
+        }
+
+        m_channels = (iio_channel **)calloc(m_channel_count, sizeof * m_channels );
+
+        for (int i = 0; i < m_channel_count; ++i)
+        {
+            struct iio_channel *chn = iio_device_get_channel(m_dev, i);
+            if (iio_channel_is_scan_element(chn))
+            {
+                m_channels[i] = chn;
+            }
+        }
```

Comment:
in this case a for loop is not alright, m_channel_count is the number of all channels which can be read with the buffer. so variable i should be incremented only if the channel is added to the list
the code should be changed to something like this:
```
i = 0;
while (i < m_scan_channel_count)
{
            struct iio_channel *chn = iio_device_get_channel(m_dev, i);
            if (iio_channel_is_scan_element(chn))
            {
                m_channels[i++] = chn;
            }
}
```